warehouse/query-core/src/main/java/datawave/query/jexl/visitors/UniqueExpressionTermsVisitor.java

```diff
@@ -0,0 +1,199 @@
+            // 5. If only one child remains, replace the parent's child with the unique child.
+            // Pull up the unique child node such that
+            // PARENT -> CURRENT -> {EQ_Child_0, EQ_Child_0}
+            // PARENT -> EQ_Child_0
+            replaceChild(node.jjtGetParent(), node, uniqueChildren.get(0));
+            
+            // if (FOO == 'bar') unwrap it such that FOO == 'bar'
+            // call to unwrap will only execute if the node is wrapped in a reference expression (parens)
+            unwrap(node, uniqueChildren.get(0));
+        }
+    }
+    
+    /**
+     * A node must contain at least 2 EQ nodes to be evaluated for duplicate terms.
+     *
+     * @param node
+     *            - an ASTOrNode or ASTAndNode
+     * @return - true if node has two or more EQNodes
+     */
+    private boolean hasMinimumRequiredEqNodes(JexlNode node) {
+        int minRequiredEqNodes = 2;
+        int curEqNodes = 0;
+        for (int ii = 0; ii < node.jjtGetNumChildren(); ii++) {
+            int nodeId = id(node.jjtGetChild(ii));
+            if (nodeId == ParserTreeConstants.JJTEQNODE) {
+                curEqNodes++;
+                if (curEqNodes == minRequiredEqNodes) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    
+    /**
+     * Removes duplicate child EQ nodes while preserving other child nodes. Order of children is also preserved.
+     *
+     * @param node
+     *            - an ASTOrNode or ASTAndNode
+     * @return - List of children with duplicate EQNodes removed
+     */
+    private List<JexlNode> removeDuplicateEqNodes(JexlNode node) {
+        Set<String> eqNodeKeys = new HashSet<>();
+        List<JexlNode> children = new ArrayList<>();
+        for (int ii = 0; ii < node.jjtGetNumChildren(); ii++) {
+            JexlNode child = node.jjtGetChild(ii);
+            if (id(child) == ParserTreeConstants.JJTEQNODE) {
+                
+                // Build key for EQ node to determine if we've seen it before
+                String key = buildKeyForNode(child);
+                
+                if (eqNodeKeys.add(key)) {
+                    // Add the EQ node if we have not seen it before
+                    children.add(child);
+                } else {
+                    // Increment duplicate count
+                    this.duplicates++;
+                }
+            } else {
+                // Retain all the other nodes, preserving order.
+                children.add(child);
+            }
+        }
+        return children;
+    }
+    
+    // Builds a simple key for this EQ node from the field name and the identifier
+    // For example (FOO == 'bar') would return a key of "FOO.bar"
+    private String buildKeyForNode(JexlNode node) {
+        String fieldName = JexlASTHelper.getIdentifier(node);
+        Object identifier = JexlASTHelper.getLiteralValue(node);
+        return fieldName + "." + identifier;
```

Comment:
more generally should this be reflective on an impl on equals/hashcode of JexlNode?

Reply:
The built in hashCode() method doesn't work, in the future a more robust equality method could be added to the JexlNodes class (static call to nodesEqual(node 1, node 2)).